revokefs/writer.c

```diff
@@ -0,0 +1,876 @@
+
+  if (unlinkat (basefd, path, AT_REMOVEDIR) == -1)
+    response->result = -errno;
+  else
+    response->result = 0;
+
+  return 0;
+}
+
+int
+request_rmdir (int writer_socket, const char *path)
+{
+  return request_path (writer_socket, REVOKE_FS_RMDIR, path);
+}
+
+static ssize_t
+handle_unlink (RevokefsRequest *request,
+              gsize data_size,
+              RevokefsResponse *response)
+{
+  g_autofree char *path = get_valid_path (request->data, data_size);
+
+  if (unlinkat (basefd, path, 0) == -1)
+    response->result = -errno;
+  else
+    response->result = 0;
+
+  return 0;
+}
+
+int
+request_unlink (int writer_socket, const char *path)
+{
+  return request_path (writer_socket, REVOKE_FS_UNLINK, path);
+}
+
+static ssize_t
+handle_symlink (RevokefsRequest *request,
+                gsize data_size,
+                RevokefsResponse *response)
+{
+  g_autofree char *from = NULL;
+  g_autofree char *to = NULL;
+
+  /* from doesn't have to be a valid path, it can be absolute or whatever */
+  get_any_path_and_valid_path (request, data_size,  &from, &to);
+
+  if (symlinkat (from, basefd, to) == -1)
+    response->result = -errno;
+  else
+    response->result = 0;
+
+  return 0;
+}
+
+int
+request_symlink (int writer_socket, const char *from, const char *to)
+{
+  return request_path_path (writer_socket, REVOKE_FS_SYMLINK, from, to);
+}
+
+static ssize_t
+handle_link (RevokefsRequest *request,
+             gsize data_size,
+             RevokefsResponse *response)
+{
+  g_autofree char *from = NULL;
+  g_autofree char *to = NULL;
+
+  get_valid_2path (request, data_size,  &from, &to);
+
+  if (linkat (basefd, from, basefd, to, 0) == -1)
+    response->result = -errno;
+  else
+    response->result = 0;
+
+  return 0;
+}
+
+int
+request_link (int writer_socket, const char *from, const char *to)
+{
+  return request_path_path (writer_socket, REVOKE_FS_LINK, from, to);
+}
+
+static ssize_t
+handle_rename (RevokefsRequest *request,
+             gsize data_size,
+             RevokefsResponse *response)
+{
+  g_autofree char *from = NULL;
+  g_autofree char *to = NULL;
+
+  get_valid_2path (request, data_size,  &from, &to);
+
+  if (renameat (basefd, from, basefd, to) == -1)
+    response->result = -errno;
+  else
+    response->result = 0;
+
+  return 0;
+}
+
+int
+request_rename (int writer_socket, const char *from, const char *to)
+{
+  return request_path_path (writer_socket, REVOKE_FS_RENAME, from, to);
+}
+
+static ssize_t
+handle_chmod (RevokefsRequest *request,
+              gsize data_size,
+              RevokefsResponse *response)
+{
+  g_autofree char *path = get_valid_path (request->data, data_size);
+  int mode = request->arg1;
+
+  /* Note we can't use AT_SYMLINK_NOFOLLOW yet;
+   * https://marc.info/?l=linux-kernel&m=148830147803162&w=2
+   * https://marc.info/?l=linux-fsdevel&m=149193779929561&w=2
+   */
+  if (fchmodat (basefd, path, mask_mode (mode), 0) != 0)
```

Comment:
I wonder if this is a problem. The user is able to create symlinks with any target value, and if fchmodat follows symlinks it could then use that to trick the writer to change the permissions of that file, which might not necessarily be inside the basedir.

Of course, we *do* limit the modes you can set, so its not a *huge* problem. Still, it seems iffy. We should maybe fail if the destination is a symlink instead.

Reply:
(To make it atomic we would need to openat() O_NOFOLLOW, and then fchmod())

Reply:
Hmm, i looked at this, but there is no atomic way to fix this. We can't use openat + fchmod, because if the current permissions is such that you can't open the file you can't use then open+fchmod to fix it.

However, fuse itself does resolve symlinks before calling chmod() in the fuse implementation, so this might be fine anyway.